### ISSUE TITLE
Author response (Usenix security)

### ISSUE THREAD

**Issue by praveenabt** [networked-systems-iith/AdaFlow#8]
We thank the reviewers for their constructive and insightful comments. We \textit{stress} that the focus of the AdaFlow system is on (i) extracting and exporting quality telemetry features (in the data plane) for a wide range of malicious flow detection, (ii) designing efficient \textit{push model-based} cache to minimize control- and data-plane communication overheads without impacting detection accuracy. One can further optimize and fine-tune ML models in the literature to improve results. To understand the benefits of AdaFlow, we pick CIC-IDS2017 with 14 different attacks as a reference dataset — evaluating AdaFlow on other datasets and different ML models is our ongoing and future work.   

\textbf{[B, C, D, E] Generality}: We agree that our first contribution statement should be rephrased. Being general, we meant that the cache primitive is designed to capture flow features (11) essential to detect a (relatively) wide range of flow-based network attacks (e.g., attacks in CIC-IDS, P2P App Fingerprinting, Covert Channel detection), but not those attacks where flow features are not recommended. Most related works (mentioned in Table 4) use a subset of these 11 features. By providing all features simultaneously, we need not modify data-plane logic for collecting attack-specific flow features like in FlowLens and NetBeacon.

\textbf{[D] Comparison with [G]}: We compare our design with current systems that efficiently extract flow features and classify flows by leveraging Tofino switch capabilities. As mentioned earlier, using fine-tuned ML models like Autoencoders in [G] in the control plane could be explored. We will list the works suggested by [D] in the revised version. We, however, did mention Tofino-based FastFE in Table-4, which builds upon [G].

\textbf{[B] Cache contention}: We picked CIC-IDS2017 dataset as it had about 14 different types of malicious flow attacks. We agree that more traffic \textit{might} change the results – we hypothesize that more traffic would impact precision and bandwidth overhead, but not recall as the cache design prioritizes malicious flows – we can get results by camera ready.

\textbf{[A, B, C, D, E] Experiments}: We did not know that CIC-IDS2017 is flawed. We looked at the references [D] provided and noted that most metrics were affected in the case of Web-Attacks, Infiltration, and Bots.  We will re-evaluate on a fixed CIC-IDS2017 version without design changes and get the results ready by the camera-ready version.

We agree that to make a better case in terms of accuracy and generality, AdaFlow should be evaluated on different datasets – P2P App Fingerprinting, Covert Channel, and DDoS detection. Relative to NetBeacon, we got good recall (directly from the online decision tree) for the first two datasets (0.83/0.85, 0.87/0.92, 0.89/0.94 – covert channel – 30, 60, and 90 hash table entries). We will conduct experiments to evaluate FPR, FNR, malicious flow loss, packet recirculations, and ROC curves and complete these experiments by the camera-ready deadline.

\textbf{[A, C, D] ML phases}: We use 75% of flows (of all PCAPs) as the training set on which we engineer the 11 features and derive importance scores of each feature for each attack type. Rest 15% formed the validation set, and 10% formed the test set – we evaluated AdaFlow with this 25% of flows (validation+test), hence \textit{no snooping bias}. We repeated our experiments five times and observed similar results.

\textbf{[E] Decision tree, distributed attacks, and generality}: Due to switch data-plane constraints on per-packet operations, it is infeasible to run complex and accurate classifiers. We envision such ML  models run servers, whereas simple DT-based online classifier retains ideal recall. Due to simplicity, DT might misclassify an evicted flow, so we \textit{fall back to} complex ML models for improving accuracy.  Currently, AdaFlow is designed to detect attacks based on flow-level features. We thank the reviewer for bringing distributed attacks that require aggregation of features \textit{across flows}. AdaFlow \textit{might} miss exporting features of flows in a distributed attack due to collisions with residing malicious flow. Despite its low probability, AdaFlow is incomplete and requires careful study. In the paper, we compared AdaFlow with attack-specific sketches. A \textit{set of sketches} covering a wide range of attacks \textit{may} provide the same features as AdaFlow, and this approach has its own design and implementation challenges – this comparison is not made in the submitted version.
